File: 2015/day05.py
VOWELS = set("aeiou")


def part1(s: str):
    vowel_count, twice = s[0] in VOWELS, False
    for i in range(len(s) - 1):
        ss = s[i : i + 2]
        if ss in {"ab", "cd", "pq", "xy"}:
            vowel_count = 0
            break
        if not twice and ss[0] == ss[1]:
            twice = True
        if ss[1] in VOWELS:
            vowel_count += 1
    return vowel_count >= 3 and twice


def part2(s: str):
    for i in range(len(s) - 3):
        for j in range(i + 2, len(s) - 1):
            if s[i : i + 2] == s[j : j + 2]:
                for i in range(len(s) - 2):
                    if s[i] == s[i + 2]:
                        return True
    return False


with open("in/d5.txt") as f:
    total1 = total2 = 0
    for s in f:
        total1 += part1(s)
        total2 += part2(s)
    print("Part 1:", total1)
    print("Part 2:", total2)

File: 2025/Solutions/Day05.cs
using System.Collections.Immutable;

namespace aoc.Solutions
{
    public class Day05
    {
        record Range(ulong Start, ulong End);
        public static void Solve()
        {
            var (ranges, ingredients) = new Func<(ImmutableArray<Range>, ImmutableArray<ulong>)>(() =>
            {
                string data = File.ReadAllText("in/d05.txt").Replace("\r\n", "\n").TrimEnd();
                string[] range_map = data.Split("\n\n");

                var ranges = range_map[0]
                    .Split("\n")
                    .Select(row =>
                    {
                        var res = row.Split("-").Select(x => ulong.Parse(x)).ToArray();
                        return new Range(res[0], res[1]);
                    })
                    .ToImmutableArray();

                var ingredients = range_map[1]
                    .Split("\n")
                    .Select(x => ulong.Parse(x))
                    .ToImmutableArray();

                return (ranges, ingredients);
            })();

            Console.WriteLine($"Part 1: {Part1(ranges, ingredients)}");
            Console.WriteLine($"Part 2: {Part2(ranges)}");
        }
        static ulong Part1(ImmutableArray<Range> ranges, ImmutableArray<ulong> ingredients)
        {
            ulong freshness = 0;

            foreach (var ingredient in ingredients)
            {
                foreach (var range in ranges)
                {
                    if (range.Start <= ingredient && ingredient <= range.End)
                    {
                        freshness += 1;
                        break;
                    }
                }
            }


            return freshness;
        }

        static ulong Part2(ImmutableArray<Range> ranges)
        {
            ulong freshness = 0;

            var a = ranges.ToList();

            while (true)
            {
                bool[] flags = new bool[a.Count];
                List<Range> b = [];

                for (int i = 0; i < a.Count; i++)
                {
                    if (flags[i]) continue;

                    var (start, end) = a[i];
                    for (int j = i + 1; j < a.Count; j++)
                    {
                        if (flags[j]) continue;

                        var (_start, _end) = a[j];
                        if (start <= _start && _start <= end)
                        {
                            end = ulong.Max(_end, end);
                            flags[j] = true;
                        }
                        if (start <= _end && _end <= end)
                        {
                            start = ulong.Min(_start, start);
                            flags[j] = true;
                        }
                    }
                    b.Add(new(start, end));
                }

                if (flags.All(x => !x))
                {
                    a = b;
                    break;
                }
                a = b;
                a.Reverse(); // THIS SOLVES IT... Yay bug somewhere :D
            }

            foreach (var range in a)
            {
                Console.WriteLine($"{range.Start},{range.End}");
                freshness += range.End - range.Start + 1;
            }

            return freshness;
        }
    }
}